### PR TITLE
fix(userstatus): Track message timestamp too

### LINKED ISSUE #40558: [Bug]: User status widget shows incorrect relative timestamp
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

The status widget for the dashboard doesn't reflect when a user set the status. It's some kind of timestamp of their last online presence, or when a client connected.

### Steps to reproduce

1. Set a status as user A
2. Open the dashboard as user B
3. Enable the status widget
4. Look at A's status

### Expected behavior

A's status and the time relative to setting the status

### Installation method

None

### Nextcloud Server version

25

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by ChristophWurst**
* Resolves: https://github.com/nextcloud/server/issues/40558

## Summary

Add new database column to track the message change date.

The fix requires a database schema change and data migration. I would vote against backporting.

## How to test

1) Set up user A
2) Set up user B
3) Log in as A
4) Log in as B in a private window
5) Set your status to DND and the status message to "test" as user A
6) Wait 2 minutes
7) Set your status to online as user A (**do not** use the *Save status message* button. Online status change doesn't need a confirmation)
7) Open the dashboard and recent statuses as user B

master: User A shows "test, seconds ago"
this branch: User A shows "test, 2 minutes ago"

## Screenshots

| Before | After |
|--------|--------|
| Bildschirmfoto vom 2023-09-26 14-31-52 |  Bildschirmfoto vom 2023-09-26 14-31-22 | 

## TODO

- [x] Adjust schema
- [x] Adjust db layer
- [x] Adjust dashboard widget
- [x] Test
- [x] Update branch after https://github.com/nextcloud/server/pull/40614

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by Antreesy — COMMENTED** [review]
:seal: 

Tested with my 0-PHP-knowledge, works fine for custom statuses.
Should it also work for predefined ones, like "Commuting"?

Also noted:
- upgrade to 1.8.1 works smoothly, status is saved
- downgrade to 1.8.0 breaks the instance, but that should be expected, i guess

**Comment by nickvergessen** [comment]
> downgrade to 1.8.0 breaks the instance, but that should be expected, i guess

Unfortunately yes. YOu can manually delete the column from the database table (but that will break on updating, unless you also remove the row in oc_migrations table, so I personally would prefer to reinstall)

**Comment by nickvergessen** [comment]
> Should it also work for predefined ones, like "Commuting"?

When selecting them I guess yes. When they fade (because clearAt was reached) and you turn into "Away/Online" I guess also yes?

**Comment by ChristophWurst** [comment]
> When they fade

Right! Forgot to test those.

When the predefined statuses are selected they should definitely update the status message timestamp.

**Comment by ChristophWurst** [comment]
> > When they fade
> 
> Right! Forgot to test those.

Going from a preset status (or custom status) to *no status* erases the database entry. There is no record of the change on the recent statuses widget. I would say this is the expected behavior, because there is no recent status to list.

**Comment by nickvergessen** [comment]
> Going from a preset status (or custom status) to no status erases the database entry.

Right, then that's how it is

**Comment by nickvergessen** [comment]
Restoring from backup status (e.g. custom status > join a call for IN_CALL hardcoded one > leave call again) should also update the timestamp? Or not?

**Comment by ChristophWurst** [comment]
Phew. It's getting complicated.

If we update the status message timestamp when an automated status is entered, it should also update when that status ends and the previous status is restored from backup. Con: status shows as updated at restore when the user didn't update it themselves.

1) "Working remotely, 2h ago"
2) "In a call, 28 min ago"
3) "Working remotely, 4m ago"

Alternatively, we can not update the timestamp for entering/leaving an automated status. Pro/con: automated statuses don't show on the recent statuses widget.

1) "Working remotely, 4h ago"

@jancborchardt are you okay with the first approach?

**Comment by ChristophWurst** [comment]
Integration test suite added so we can be sure the strange edge cases all work as expected :relaxed:

**Comment by ChristophWurst** [comment]
@Antreesy @nickvergessen if you are okay with the current solution we can integrate this (and unblock https://github.com/nextcloud/server/issues/40559). If we decide that automated/backup statuses should update the timestamp as well we can still change that behavior later on :+1:

**Comment by nickvergessen** [comment]
> if you are okay with the current solution we can integrate this (and unblock https://github.com/nextcloud/server/issues/40559). If we decide that automated/backup statuses should update the timestamp as well we can still change that behavior later on 👍

Agreed

File: apps/user_status/composer/composer/autoload_classmap.php [hunk 1]
```diff
@@ -30,7 +30,13 @@
     'OCA\\UserStatus\\Migration\\Version0001Date20200602134824' => $baseDir . '/../lib/Migration/Version0001Date20200602134824.php',
     'OCA\\UserStatus\\Migration\\Version0002Date20200902144824' => $baseDir . '/../lib/Migration/Version0002Date20200902144824.php',
     'OCA\\UserStatus\\Migration\\Version1000Date20201111130204' => $baseDir . '/../lib/Migration/Version1000Date20201111130204.php',
+<<<<<<< HEAD
```
Comment (nickvergessen) [inline apps/user_status/composer/composer/autoload_classmap.php:33]: please retry

Comment (ChristophWurst) [reply apps/user_status/composer/composer/autoload_classmap.php:33]: https://classicprogrammerpaintings.com/post/142586036029/junior-programmer-learns-git-rebase

**Review by nickvergessen — APPROVED** [review]

**Review by Antreesy — APPROVED** [review]

**Review by miaulalala — APPROVED** [review]
